### DISCUSSION TITLE
Help: Don't know how interpret this AOT problem

### DISCUSSION THREAD

**Discussion by SwingWren** [applejag/Newtonsoft.Json-for-Unity#52] [Q&A] [answered]
Hello! :). I get this error "System.ArgumentException: Member 'System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte)' " during compilation.
What I should put in the link.xml ? I was able to solve other problems but I cannot figure this one out.

## Expected behavior

To compile without an error;

## Actual behavior

Crash during compilation 
"System.ArgumentException: Member 'System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte)' "

## Steps to reproduce
- New project
- Import `jillejr.newtonsoft.json-for-unity` via UPM
- Add following script to scene:

```csharp
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using UnityEngine;

public class NewBehaviourScript : MonoBehaviour {
  // Start is called before the first frame update
  void Start() {
    JObject item = new JObject();
    item["oneKey"] = JObject.FromObject(new RandomItem {anyValue = 1});
    string indentedItem = item.ToString(Formatting.Indented);
    Debug.Log(indentedItem);
  }
}

public class RandomItem {
  public int anyValue;
}
```

- Build using  IL2CPP with High Stripping level

## Details

Host machine OS running Unity Editor: Windows

Unity build target Windows/WebGL build as High

Newtonsoft.Json-for-Unity package version 12.0.301

I was using Unity version 2019.3.12f1

## Checklist
- [x] Shutdown Unity, deleted the /Library folder, opened project again in Unity, and problem still remains.
- [x] Checked to be using latest version of the package.

**Comment by applejag** [comment]
Hi @SwingWren, interesting scenario you've found where as you say it doesn't even compile. I've sadly not done too much testing with High Stripping Level, but I'll investigate this and hopefully we can come to a solution ASAP! :)

**Comment by applejag** [comment, marked as answer]
I love this issue ❤ because the error roots so far down in the IL, and I rarely get to analyse that deep.

What's going on is that since Json .NET 12.0.3 it actually includes the NullableAttribute and NullableContextAttribute for C# 8's nullable ref types.

The High stripping level for some reason removes the `NullableAttribute:.ctor(byte)` constructor, even though it's clearly referenced in the `JObject.set_Item` method (used in the `item["oneKey"] =` line).

![image](https://user-images.githubusercontent.com/2477952/81505049-09f24c00-92ed-11ea-9ffa-cf8b7a7ae583.png)

That IL declaration is the equivalent to the following C# code:

```cs
public void set_Item(string propertyName, [Nullable((byte)2)] JToken value) {}
```

## Workaround

This is what you actually asked for. Adding the `NullableAttribute` to your `link.xml` will make it all safe to use.

It seems like it's only necessary to specify the `.ctor(byte)` constructor to make it build, but to be on the safe side you may declare the entire `NullableAttribute` and the accompanying `NullableContextAttribute` to be preserved by adding the following:

```xml
<linker>
  <assembly fullname="Newtonsoft.Json">
    <type fullname="System.Runtime.CompilerServices.NullableAttribute"/>
    <type fullname="System.Runtime.CompilerServices.NullableContextAttribute"/>
  </assembly>
</linker>
```

And yes, it's supposed to be embedded in the Newtonsoft.Json assembly.

## Long-term fix

I'll probably just write a small thread on the Unity forums, hoping some Unity dev will see it and can perhaps fix it, as it seems like this is a problem with IL2CPP itself.

In the meantime I'll toss in the NullableAttribute into the `link.xml` that comes with Newtonsoft.Json-for-Unity so it wont occur again for 12.0.302 and higher.

---

Closing as resolved. Don't fray in reopening the issue if the problem comes back with some other type.

Cheers!
